Require newer version of llvm-openmp for xgboost and py_torch on macOS

zackgalbreath opened the pull request:
* Add v19.1.7 and v20.1.8 for llvm-openmp
* Depend on llvm-openmp v19 or newer when building xgboost or torch with OpenMP using AppleClang
* Remove py-torchtext from ml-darwin-aarch64-mps so we can re-enable the stack while debugging this build failure

zackgalbreath committed:
llvm-openmp: add v19.1.7 and v20.1.8

zackgalbreath committed:
Require newer version of llvm-openmp for xgboost and py_torch on macOS

Depend on llvm-openmp v19 or newer when building xgboost or torch with
OpenMP using AppleClang.

zackgalbreath commented:
torch still fails to build for the MPS stack, but this PR seems to fix xgboost at least:

https://gitlab.spack.io/spack/spack-packages/-/jobs/17424434

tldahlgren reviewed (dismissed):
New version sha256s confirmed for `llvm-openmp`. 

@adamjstewart Any objections to the changes to `py-torch` or `xgboost`?

zackgalbreath commented:
Here's some additional evidence that bumping the version of llvm-openmp is the right way to get xgboost building successfully again:
* https://github.com/llvm/llvm-project/issues/98046#issuecomment-2905413530
* https://trac.macports.org/ticket/70877

zackgalbreath commented:
@kwryankrattiger @eugeneswalker FYI. This PR intends to get us closer to restoring the `ml-darwin-aarch64-mps` stack.

zackgalbreath committed:
Remove py-torchtext from ml-darwin-aarch64-mps

Temporarily disable this package while we debug build errors.
This change will allow us to re-enable the rest of the stack.

zackgalbreath commented:
@kwryankrattiger @eugeneswalker the latest passing pipeline for this PR includes the `ml-darwin-aarch64-mps` stack with `py-torchtext` removed.

https://gitlab.spack.io/spack/spack-packages/-/pipelines/1160721

kwryankrattiger reviewed (approved):
LGTM, good enough to reenable the rest of this stack and we can follow up to get `pytorch-ext` reenabled later.

FYI: @adamjstewart

adamjstewart commented on repos/spack_repo/builtin/packages/py_torch/package.py:
```diff
@@ -299,7 +299,7 @@ class PyTorch(PythonPackage, CudaPackage, ROCmPackage):
     depends_on("magma+rocm", when="+magma+rocm")
     depends_on("numactl", when="+numa")
-    depends_on("llvm-openmp", when="+openmp %apple-clang")
+    depends_on("llvm-openmp@19:", when="+openmp %apple-clang")
```
Curious where these constraints come from. I've never encountered any problems building PyTorch with older OpenMP before.

zackgalbreath replied:
You're right, that might not be necessary (although it was for the xgboost fix). I'll try relaxing this constraint again in #641 while this is fresh in our minds.

adamjstewart replied:
Can you share error messages for XGBoost and/or PyTorch?

zackgalbreath replied:
Sure!

* For XGBoost it was: [`Undefined symbols for architecture arm64: "___kmpc_dispatch_deinit"`](https://gitlab.spack.io/spack/spack-packages/-/jobs/17373616#L835)
    - See [my comment here](https://github.com/spack/spack-packages/pull/619#issuecomment-3080040192) for other folks that ran into this & fixed it by version bumping OpenMP.

* For Torch it was:
[`error: always_inline function 'vmul_f16' requires target feature 'fullfp16', but would be inlined into function 'fp16_gemv_trans_fp16_arith' that is compiled without support for 'fullfp16'`](https://gitlab.spack.io/spack/spack-packages/-/jobs/17424425#L4869)
    - Note that this only seemed to be a problem for `py_torch@2.3.0` (the version needed by TorchText). This issue seems to be fixed by https://github.com/spack/spack/pull/51011 .

adamjstewart replied:
* XGBoost: when did this start, after an OS update? Maybe we could condition based on compiler
* PyTorch: yeah, let's remove if this was a bug in Spack's compiler wrapper, glad this was fixed!
